### PR TITLE
Remove most use of "Bogus" from Microsoft.CSharp and fix remaining

### LINKED ISSUE #22534: Microsoft.CSharp does not report CS1545 or CS1546
If a type has property with parameters that is not its indexer property, then attempting to index it in C# will cause either CS1545 if it has both a getter  and setter ("Property, indexer, or event 'XXX' is not supported by the language; try directly calling accessor methods 'get_XXX' or 'set_XXX'") or CS1546 if it has only one of those ("Property, indexer, or event 'XXX' is not supported by the language; try directly calling accessor method 'get_XXX'").

If the same thing happens with dynamic code, it throws an exception equivalent to a less helpful (and slightly misleading) CS0570 with a message of "'TheType' is not supported by the language".

There's code to report CS1545 or CS1546, but it fails to identify the situation when that code should be run.

### PR DESCRIPTION
**Pull request by JonHanna**
* Remove concept of bogus from types and most symbols.

Since bogus types can't get as far as the dynamic binder, the only case where bogus can be true is when a non-indexer parametrized property, defined in a language where that is allowed, is encountered.

In other cases computing bogosity involves recursive checks that always hit false.

Give `PropertySymbol` a separate `Bogus` property to use where it could be set, and remove all other uses of `GetBogus` and similar.

Entails removal of `ERR_BogusType`, contributes to #22470

* Identify indexer properties correctly in `AddPropertyToSymbolTable`

`AddPropertyToSymbolTable` assumes any property with parameters is an indexer, which makes the subsequent check that only indexers can have parameters tautologous.

Check `DefaultMemberAttribute` instead.

* Remove `useMethInstead` from property symbols and consider always true.

`useMethInstead` should be set in `AddPropertyToSymbolTable` when bogus is set, but is not.

However, the only time it is looked for the only possible value is true as that is the only condition under which a symbol can be bogus, so remove it, and the path for it being false.

This now means that `ErrorCode.ERR_BindToBogusProp1` or `ErrorCode.ERR_BindToBogusProp2` will be used when appropriate (contributes to #22470).

Fixes #22534

* Tests for accessing non-indexer parameter properties

* Remove bogus check within PredefinedMembers

Predefined properties cannot be bogus.

### CODE REVIEW

**Comment by JonHanna** [comment]
Test OSX x64 Debug Build

File: src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/SymbolTable.cs [hunk 1]
```diff
@@ -1354,7 +1354,9 @@ internal void AddPredefinedPropertyToSymbolTable(AggregateSymbol type, Name prop
         private void AddPropertyToSymbolTable(PropertyInfo property, AggregateSymbol aggregate)
         {
             Name name;
-            bool isIndexer = property.GetIndexParameters() != null && property.GetIndexParameters().Length != 0;
+            bool isIndexer = property.GetIndexParameters().Length != 0
```
Comment (OmarTawfik) [inline src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/SymbolTable.cs]: Please use Any()

File: src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/SymbolTable.cs [hunk 2]
```diff
@@ -1354,7 +1354,9 @@ internal void AddPredefinedPropertyToSymbolTable(AggregateSymbol type, Name prop
...
+                             && (property.DeclaringType?.GetCustomAttribute(typeof(DefaultMemberAttribute)) as DefaultMemberAttribute)
```
Comment (OmarTawfik) [inline src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/SymbolTable.cs]: >DefaultMemberAttribute [](start = 83, length = 22)

GetCustomAttribute<DefaultMemberAttribute>()?

File: src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/SemanticChecker.cs [hunk 3]
```diff
@@ -288,46 +288,7 @@ private ACCESSERROR CheckAccessCore(Symbol symCheck, AggregateType atsCheck, Sym
 
         public bool CheckBogus(Symbol sym)
```
Comment (OmarTawfik) [inline src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/SemanticChecker.cs]: Please use this everywhere this functionality is needed. Example: GroupToArgsBinder.LookForCandidates()

File: src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/Symbols/PropertySymbol.cs [hunk 4]
```diff
@@ -31,5 +31,7 @@ public IndexerSymbol AsIndexerSymbol()
             Debug.Assert(isIndexer());
             return (IndexerSymbol)this;
         }
+
+        public bool Bogus { get; set; }
```
Comment (OmarTawfik) [inline src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/Symbols/PropertySymbol.cs]: I suggest keeping with other fields, or converting all fields to properties.
I know there is a bigger effort of refactoring these things, but we shouldn't special case this here now.

File: src/Microsoft.CSharp/tests/BindingErrors.cs [hunk 5]
```diff
@@ -221,5 +224,16 @@ public void AmbigousUnaryOp()
             dynamic d = new AmbiguousNumClass(7);
             Assert.Throws<RuntimeBinderException>(() => -d);
         }
+
+        [Fact]
+        public void LiteralDoubleCast()
```
Comment (OmarTawfik) [inline src/Microsoft.CSharp/tests/BindingErrors.cs]: Not sure how this relates to this PR. Maybe from #22533?

File: src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/SymbolTable.cs (hunk 2 above)
Comment (JonHanna) [inline src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/SymbolTable.cs]: Sorry, I don't understand this comment.

File: src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/SymbolTable.cs (hunk 1 above)
Comment (JonHanna) [inline src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/SymbolTable.cs]: Any reason to use predicateless-Any() on an array? The perf difference isn't much, but it's there, and `Length != 0` is common with arrays elsewhere in corefx.

File: src/Microsoft.CSharp/tests/BindingErrors.cs (hunk 5 above)
Comment (JonHanna) [inline src/Microsoft.CSharp/tests/BindingErrors.cs]: Yeah, something slipped while doing a bunch of rebases. Sorry.

File: src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/SemanticChecker.cs (hunk 3 above)
Comment (JonHanna) [inline src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/SemanticChecker.cs]: Done. Have also made it static as it requires no state.

File: src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/Symbols/PropertySymbol.cs (hunk 4 above)
Comment (JonHanna) [inline src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/Symbols/PropertySymbol.cs]: `isIndexer` and `AsIndexerSymbol` should ideally have been changed when getting rid of the similar methods defined elsewhere, anyway.

File: src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/SymbolTable.cs (hunk 2 above)
Comment (JonHanna) [inline src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/SymbolTable.cs]: Oh. TIL about the generic `GetCustomAttribute<>()` extension!

File: src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/SymbolTable.cs (hunk 1 above)
Comment (VSadov) [inline src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/SymbolTable.cs]: I think Length is ok when we know it is an array. "Any" is often faster with various kind of collections/enumerables, but for array we know the "Length != 0" is the simplest way to check for emptiness.

---
In reply to: [129472003](https://github.com/dotnet/corefx/pull/22537#discussion_r129472003) [](ancestors = 129472003)

File: src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/Symbols/MethodOrPropertySymbol.cs [hunk 6]
```diff
@@ -29,7 +29,6 @@ internal abstract class MethodOrPropertySymbol : ParentSymbol
         public new bool isStatic;               // Static member?
         public bool isOverride;             // Overrides an inherited member. Only valid if isVirtual is set.
         // false implies that a new vtable slot is required for this method.
-        public bool useMethInstead;         // Only valid iff isBogus == TRUE && IsPropertySymbol().
         // If this is true then tell the user to call the accessors directly.
```
Comment (VSadov) [inline src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/Symbols/MethodOrPropertySymbol.cs]: This comment refers to the code that is now gone. Also the comment above looks like after the code it is referring to. Can that be fixed too?

Comment (JonHanna) [inline src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/Symbols/MethodOrPropertySymbol.cs]: Good point. Done.

**Pull request merged**